python_modules/dagster/dagster/core/definitions/repository_definition.py

```diff
@@ -1161,3 +1172,52 @@ def source_assets_by_key(self) -> Dict[AssetKey, SourceAsset]:
     # overwritten. Therefore, we want to maintain the call-ability of repository definitions.
     def __call__(self, *args, **kwargs):
         return self
+
+
+def _process_and_validate_target(
+    schedule_or_sensor_def: Union[SensorDefinition, ScheduleDefinition],
+    coerced_graphs: Dict[str, JobDefinition],
+    pipelines_or_jobs: Dict[str, PipelineDefinition],
+    target: Union[GraphDefinition, PipelineDefinition],
+):
+    # This function modifies the state of coerced_graphs.
+    targeter = (
+        f"schedule '{schedule_or_sensor_def.name}'"
+        if isinstance(schedule_or_sensor_def, ScheduleDefinition)
+        else f"sensor '{schedule_or_sensor_def.name}'"
+    )
+    if isinstance(target, GraphDefinition):
+        if target.name not in coerced_graphs:
+            # Since this is a graph we have to coerce, is not possible to be
+            # the same definition by reference equality
+            if target.name in pipelines_or_jobs:
+                dupe_target_type = pipelines_or_jobs[target.name].target_type
+                warnings.warn(
+                    _get_error_msg_for_target_conflict(
+                        targeter, "graph", target.name, dupe_target_type
+                    )
+                )
+        elif coerced_graphs[target.name].graph != target:
+            warnings.warn(
+                _get_error_msg_for_target_conflict(targeter, "graph", target.name, "graph")
+            )
+        coerced_job = target.coerce_to_job()
+        coerced_graphs[target.name] = coerced_job
+        pipelines_or_jobs[target.name] = coerced_job
+    else:
+        if target.name in pipelines_or_jobs and pipelines_or_jobs[target.name] != target:
+            dupe_target_type = (
+                pipelines_or_jobs[target.name].target_type
+                if target.name not in coerced_graphs
+                else "graph"
+            )
+            warnings.warn(
+                _get_error_msg_for_target_conflict(
+                    targeter, target.target_type, target.name, dupe_target_type
+                )
+            )
+        pipelines_or_jobs[target.name] = target
+
+
+def _get_error_msg_for_target_conflict(targeter, target_type, target_name, dupe_target_type):
+    return f"{targeter} targets {target_type} '{target_name}', but a different {dupe_target_type} with the same name was provided. The {target_type} provided to {targeter} will override the existing {dupe_target_type}, but in Dagster 0.15.0, this will result in an error. Disambiguate between these by providing a separate name to one of them."
```

Comment:
I can't think of anything better. Every error message case gets kind of weird: `disambiguate between type1 and type2` gets weird when they are the same type: `disambiguate between the graph and the graph`, you can't do it by name either because they have the same name lol.